pkg/replication/replication-client.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"time"
+
+	replicationlib "github.com/kube-storage/spec/lib/go/replication"
+	"google.golang.org/grpc"
+)
+
+type replicationClient struct {
+	client  replicationlib.ControllerClient
+	timeout time.Duration
+}
+
+// VolumeReplication holds the methods requried for volume replication.
+type VolumeReplication interface {
+	// EnableVolumeReplication RPC call to enable the volume replication.
+	EnableVolumeReplication(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.EnableVolumeReplicationResponse, error)
+	// DisableVolumeReplication RPC call to disable the volume replication.
+	DisableVolumeReplication(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.DisableVolumeReplicationResponse, error)
+	// PromoteVolume RPC call to promote the volume.
+	PromoteVolume(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.
+		PromoteVolumeResponse, error)
+	// DemoteVolume RPC call to demote the volume.
+	DemoteVolume(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.
+		DemoteVolumeResponse, error)
+	// ResyncVolume RPC call to resync the volume.
+	ResyncVolume(ctx context.Context, volumeID string, secrets, parameters map[string]string) (*replicationlib.
+		ResyncVolumeResponse, error)
+}
+
+// NewReplicationClient returns VolumeReplication interface which has the RPC
+// calls for replication.
+func NewReplicationClient(cc *grpc.ClientConn, timeout time.Duration) VolumeReplication {
```

Comment:
If this is per volume (which would be short lived as a volume is handled in one iteration of the reconcile function) we can possibly enhance the New by adding the volID and secrets, such that other calls are not passing this in repetitively.

If this is once per ReplicationClass, we may not want to retain secrets in here for the controller lifetime.

These can be improved later, as it stands it makes code more elegant rather than creating structures everywhere before calling the gRPC.

Reply:
good for enhancements  :+1:  will keep this in mind.